src/library/crate/cratestorage.cpp

```diff
@@ -0,0 +1,665 @@
+                    CRATE_TABLE));
+    if (query.execPrepared() && query.next()) {
+        uint result = query.fieldValue(0).toUInt();
+        DEBUG_ASSERT(!query.next());
+        return result;
+    } else {
+        return 0;
+    }
+}
+
+
+bool CrateStorage::readCrateById(CrateId id, Crate* pCrate) const {
+    FwdSqlQuery query(m_database, QString(
+            "SELECT * FROM %1 WHERE %2=:id").arg(
+                    CRATE_TABLE,
+                    CRATETABLE_ID));
+    query.bindValue(":id", id);
+    if (query.execPrepared()) {
+        CrateSelectIterator crates(query);
+        if ((pCrate != nullptr) ? crates.readNext(pCrate) : crates.next()) {
+            DEBUG_ASSERT_AND_HANDLE(!crates.next()) {
+                qWarning() << "Ambiguous crate id:" << id;
+            }
+            return true;
+        } else {
+            qWarning() << "Crate not found by id:" << id;
+        }
+    }
+    return false;
+}
+
+
+bool CrateStorage::readCrateByName(const QString& name, Crate* pCrate) const {
+    FwdSqlQuery query(m_database, QString(
+            "SELECT * FROM %1 WHERE %2=:name").arg(
+                    CRATE_TABLE,
+                    CRATETABLE_NAME));
+    query.bindValue(":name", name);
+    if (query.execPrepared()) {
+        CrateSelectIterator crates(query);
+        if ((pCrate != nullptr) ? crates.readNext(pCrate) : crates.next()) {
+            DEBUG_ASSERT_AND_HANDLE(!crates.next()) {
+                qWarning() << "Ambiguous crate name:" << name;
+            }
+            return true;
+        } else {
+            qDebug() << "Crate not found by name:" << name;
+        }
+    }
+    return false;
+}
+
+
+CrateSelectIterator CrateStorage::selectCrates() const {
+    FwdSqlQuery query(m_database, QString(
+            "SELECT * FROM %1 ORDER BY %2 COLLATE %3").arg(
+            CRATE_TABLE,
+            CRATETABLE_NAME,
+            DbConnection::kStringCollationFunc));
+
+    if (query.execPrepared()) {
+        return CrateSelectIterator(query);
+    } else {
+        return CrateSelectIterator();
+    }
+}
+
+
+CrateSelectIterator CrateStorage::selectCratesByIds(
+        const QString& subselectForCrateIds,
+        SqlSubselectMode subselectMode) const {
+    QString subselectPrefix;
+    switch (subselectMode) {
+    case SQL_SUBSELECT_IN:
+        if (subselectForCrateIds.isEmpty()) {
+            // edge case: no crates
+            return CrateSelectIterator();
+        }
+        subselectPrefix = "IN";
+        break;
+    case SQL_SUBSELECT_NOT_IN:
+        if (subselectForCrateIds.isEmpty()) {
+            // edge case: all crates
+            return selectCrates();
+        }
+        subselectPrefix = "NOT IN";
+        break;
+    }
+    DEBUG_ASSERT(!subselectPrefix.isEmpty());
+    DEBUG_ASSERT(!subselectForCrateIds.isEmpty());
+
+    FwdSqlQuery query(m_database, QString(
+            "SELECT * FROM %1 WHERE %2 %3 (%4) ORDER BY %5 COLLATE %6").arg(
```

Comment:
>The entire subselect expression is provided by the caller and is not a
simple value. I don't see a use case for named parameters here?!

OK, than it makes no sense.